src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -67,6 +67,15 @@ datalog::Predicate CanActAsToDLIR(const CanActAs& can_act_as) {
 }
 }  // namespace
 
+void LoweringToDatalogPass::UpdateCanSayDepth(std::string_view predicate_name,
+                                              uint64_t depth) {
+  if (can_say_depth_.find(predicate_name) == can_say_depth_.end()) {
+    can_say_depth_.insert({predicate_name, depth});
+  } else if (depth > can_say_depth_.find(predicate_name)->second) {
+    can_say_depth_.find(predicate_name)->second = depth;
+  }
+}
```

Comment:
`can_say_depth` is a map, so this can be made even simpler. On C++ maps, `std::insert` returns a result indicating if the result successfully occurred or whether there was a key collision on insertion. If there is a key collision, it does not perform the insertion over the key, instead returning an iterator to the existing entry.

This means that you can rewrite this like so:

```C++
auto insert_result = can_say_depth.insert({predicate_name, depth});
if (!insert_result.second) {
  insert_result.first->second = std::max(depth, insert_result.first->second);
}
```

(although some temporaries to

Reply:
That is a very good optimization I was not aware of. Thanks Mark!